src/pegen/tokenizer.py

```diff
@@ -1,27 +1,121 @@
 import token
-import tokenize
-from typing import Dict, Iterator, List
+from abc import abstractmethod
+from typing import Dict, Generic, Iterator, List, Protocol, Sequence, TypeVar
 
 Mark = int  # NewType('Mark', int)
 
-exact_token_types = token.EXACT_TOKEN_TYPES
 
+TokenType = TypeVar(
+    "TokenType",
+)
 
-def shorttok(tok: tokenize.TokenInfo) -> str:
-    return "%-25.25s" % f"{tok.start[0]}.{tok.start[1]}: {token.tok_name[tok.type]}:{tok.string!r}"
 
+class TokenInfo(Protocol[TokenType]):
+    """Expected interface for element yielded by the token generator."""
 
-class Tokenizer:
-    """Caching wrapper for the tokenize module.
+    type: TokenType
+    string: str
+    start: tuple[int, int]
+    end: tuple[int, int]
+    line: str
+
+
+class TokenTypes(Generic[TokenType]):
```

Comment:
the annotations of abstract methods don't result in an exception when they are not implemented as the class doesn't derive from `ABC`. if that is intended, it should be mentioned, or the methods can raise a `NotImplementedError`.

Reply:
Fixed

Reply:
Fixed in latest commit